File: Orchestrator/src/agents/BabyAGIAgent.js
const { OpenAI } = require('langchain/llms/openai');
const BaseAgent = require('../core/BaseAgent');

// Simple in-memory vector store
class SimpleVectorStore {
  constructor() {
    this.vectors = new Map();
  }

  async upsert(id, vector, metadata) {
    this.vectors.set(id, { vector, metadata });
  }

  async query(queryVector, topK = 5) {
    // Simple cosine similarity search
    const scores = Array.from(this.vectors.entries()).map(([id, entry]) => ({
      id,
      score: this.cosineSimilarity(queryVector, entry.vector),
      metadata: entry.metadata
    }));

    return scores
      .sort((a, b) => b.score - a.score)
      .slice(0, topK);
  }

  cosineSimilarity(vecA, vecB) {
    const dotProduct = vecA.reduce((sum, a, i) => sum + a * vecB[i], 0);
    const magnitudeA = Math.sqrt(vecA.reduce((sum, a) => sum + a * a, 0));
    const magnitudeB = Math.sqrt(vecB.reduce((sum, b) => sum + b * b, 0));
    return dotProduct / (magnitudeA * magnitudeB);
  }

  async clear() {
    this.vectors.clear();
  }
}

class BabyAGIAgent extends BaseAgent {
  constructor(config) {
    super(config);
    
    // Core BabyAGI configuration
    this.objective = config.objective || '';
    this.initialTask = config.initialTask || '';
    this.modelName = config.modelName || 'gpt-4';
    this.maxIterations = config.maxIterations || 5;
    
    // Task management
    this.taskList = [];
    this.taskIdCounter = 1;
    this.tasksDone = [];
    this.vectorStore = new SimpleVectorStore();
  }

  async initialize() {
    // Check for required API key
    if (!this.config.apiKeys?.openai) {
      throw new Error('OpenAI API key is required but not provided in configuration');
    }

    // Initialize LLM
    this.llm = new OpenAI({
      modelName: this.modelName,
      temperature: 0.7,
      maxTokens: 2000,
      openAIApiKey: this.config.apiKeys.openai
    });
  }

  async createTask(taskDescription, dependentTaskId = null) {
    const task = {
      taskId: this.taskIdCounter++,
      taskName: taskDescription,
      dependentTaskId,
      status: 'pending',
      result: null,
      createdAt: new Date().toISOString()
    };
    
    this.taskList.push(task);
    return task;
  }

  async prioritizeTasks() {
    const taskNames = this.taskList.map(t => t.taskName).join(', ');
    const prompt = `You are an AI task prioritization system. You are reviewing the following tasks: ${taskNames}
    
Consider the objective: ${this.objective}
    
Analyze tasks for:
1. Dependency requirements
2. Complexity and time requirements
3. Value towards the objective
4. Current context and status
    
Provide a numbered list of these tasks in optimal order.`;

    const response = await this.llm.call(prompt);
    const prioritizedTaskNames = response.split('\n')
      .filter(line => line.match(/^\d+\./))
      .map(line => line.replace(/^\d+\.\s*/, '').trim());

    // Reorder task list based on prioritization
    this.taskList = prioritizedTaskNames.map(name => 
      this.taskList.find(task => task.taskName.includes(name))
    ).filter(Boolean);
  }

  async executeTask(task) {
    const context = await this.getTaskContext(task);
    
    const prompt = `You are an AI agent working on the following objective: ${this.objective}
    
Current task: ${task.taskName}
Context from similar tasks: ${context}
    
Execute this task and provide:
1. Detailed results
2. Any new tasks identified
3. Status update`;

    const response = await this.llm.call(prompt);
    
    // Parse response for results and new tasks
    const [results, newTasks] = this.parseTaskResponse(response);
    
    // Update task status
    task.status = 'completed';
    task.result = results;
    this.tasksDone.push(task);
    
    // Store results in vector database
    await this.storeResult(task, results);
    
    // Create new tasks
    for (const newTask of newTasks) {
      await this.createTask(newTask, task.taskId);
    }
    
    return results;
  }

  async getTaskContext(task) {
    const embedding = await this.getEmbedding(task.taskName);
    const results = await this.vectorStore.query(embedding, 5);
    return results.map(match => match.metadata.result).join('\n');
  }

  async storeResult(task, result) {
    const embedding = await this.getEmbedding(task.taskName);
    await this.vectorStore.upsert(
      task.taskId.toString(),
      embedding,
      {
        taskName: task.taskName,
        result: result,
        objective: this.objective
      }
    );
  }

  parseTaskResponse(response) {
    // Split response into results and new tasks
    const sections = response.split('\n\n');
    const results = sections[0];
    const newTasks = sections
      .find(s => s.toLowerCase().includes('new tasks:'))
      ?.split('\n')
      .filter(line => line.trim() && !line.toLowerCase().includes('new tasks:'))
      .map(task => task.replace(/^-\s*/, '')) || [];
    
    return [results, newTasks];
  }

  async getEmbedding(text) {
    // Simple mock embedding - in production, use OpenAI's embedding API
    return Array(1536).fill(0).map(() => Math.random());
  }

  async execute(input) {
    try {
      await this.initialize();
      
      // Create initial task if none exists
      if (this.taskList.length === 0) {
        await this.createTask(this.initialTask || input);
      }
      
      let iteration = 0;
      while (this.taskList.length > 0 && iteration < this.maxIterations) {
        // Prioritize tasks
        await this.prioritizeTasks();
        
        // Get next task
        const task = this.taskList.shift();
        
        // Execute task
        const result = await this.executeTask(task);
        
        // Store in memory
        await this.memory.saveContext(
          { input: task.taskName },
          { output: result }
        );
        
        iteration++;
      }
      
      // Return final results
      return {
        tasksCompleted: this.tasksDone.map(task => ({
          taskName: task.taskName,
          result: task.result
        })),
        remainingTasks: this.taskList.length,
        iterations: iteration
      };
      
    } catch (error) {
      console.error('Error executing BabyAGI agent:', error);
      throw error;
    }
  }

  async cleanup() {
    await this.vectorStore.clear();
    if (this.memory) {
      await this.memory.cleanup();
    }
  }
}

// Handler function for BabyAGI agent
const executeBabyAGIAgent = async (input, config) => {
  const agent = new BabyAGIAgent(config);
  const result = await agent.execute(input);
  await agent.cleanup();
  return result;
};

module.exports = {
  BabyAGIAgent,
  executeBabyAGIAgent
}; 

File: python-llm-service/src/agents/babyagi_agent.py
from typing import Dict, Any, List
import os
from dotenv import load_dotenv
from langchain import LLMChain, OpenAI, PromptTemplate
from langchain.embeddings import OpenAIEmbeddings
from langchain.vectorstores import Chroma
from langchain.docstore import InMemoryDocstore
from collections import deque
import numpy as np
import asyncio
from .base_agent import BaseAgent

load_dotenv()

class BabyAGIAgent(BaseAgent):
    def __init__(self, config: Dict[str, Any]):
        """Initialize BabyAGI agent with vector store and task chains."""
        super().__init__(config)
        
        # Initialize embeddings and vector store
        self.embeddings = OpenAIEmbeddings()
        self.vectorstore = Chroma(
            collection_name="tasks",
            embedding_function=self.embeddings
        )
        
        # Initialize task chains
        self.task_creation_chain = self._create_task_creation_chain()
        self.task_prioritization_chain = self._create_task_prioritization_chain()
        self.execution_chain = self._create_execution_chain()
        
        # Initialize task lists
        self.task_list = []
        self.completed_tasks = []
        self.task_id_counter = 0

    def _create_task_creation_chain(self) -> LLMChain:
        """Create the chain for generating new tasks."""
        prompt = PromptTemplate(
            input_variables=["result", "task_description", "incomplete_tasks", "completed_tasks"],
            template="""You are an AI task creation agent. Based on the result of the last task and the current task list, create new tasks to be completed that will help achieve the original objective.

Last task result: {result}
Current task: {task_description}
Incomplete tasks: {incomplete_tasks}
Completed tasks: {completed_tasks}

Create new tasks that are needed. Return only the new tasks, one per line."""
        )
        return LLMChain(llm=self.llm, prompt=prompt)

    def _create_task_prioritization_chain(self) -> LLMChain:
        """Create the chain for prioritizing tasks."""
        prompt = PromptTemplate(
            input_variables=["task_names"],
            template="""You are an AI task prioritization agent. You have the following tasks:

{task_names}

Prioritize these tasks based on:
1. Dependencies between tasks
2. Complexity and time required
3. Impact on achieving the overall objective

Return the tasks in prioritized order, one per line."""
        )
        return LLMChain(llm=self.llm, prompt=prompt)

    def _create_execution_chain(self) -> LLMChain:
        """Create the chain for executing tasks."""
        prompt = PromptTemplate(
            input_variables=["objective", "task"],
            template="""You are an AI task execution agent. Your objective is: {objective}

Your current task is: {task}

Execute this task and provide a detailed response that can be used for future tasks.
Include any relevant information, insights, or conclusions."""
        )
        return LLMChain(llm=self.llm, prompt=prompt)

    async def execute(self, user_input: str) -> Dict[str, Any]:
        """Execute the BabyAGI agent's main loop."""
        objective = user_input
        results = []
        
        # Create initial task
        self.task_list.append({
            "id": self.task_id_counter,
            "name": "Analyze objective and create initial tasks",
            "description": f"Analyze the objective: {objective} and break it down into initial tasks"
        })
        self.task_id_counter += 1
        
        # Main loop
        while self.task_list and len(results) < self.config.get("max_iterations", 5):
            # Get next task
            current_task = self.task_list.pop(0)
            
            # Execute task
            execution_result = await self.execution_chain.arun(
                objective=objective,
                task=current_task["description"]
            )
            
            # Store result
            results.append({
                "task": current_task,
                "result": execution_result
            })
            self.completed_tasks.append(current_task)
            
            # Create new tasks
            new_tasks = await self.task_creation_chain.arun(
                result=execution_result,
                task_description=current_task["description"],
                incomplete_tasks="\n".join(t["name"] for t in self.task_list),
                completed_tasks="\n".join(t["name"] for t in self.completed_tasks)
            )
            
            # Add new tasks
            for task in new_tasks.split("\n"):
                task = task.strip()
                if task:
                    self.task_list.append({
                        "id": self.task_id_counter,
                        "name": task,
                        "description": task
                    })
                    self.task_id_counter += 1
            
            # Prioritize tasks
            if self.task_list:
                prioritized_tasks = await self.task_prioritization_chain.arun(
                    task_names="\n".join(t["name"] for t in self.task_list)
                )
                
                # Reorder task list based on prioritization
                priority_dict = {task.strip(): i for i, task in enumerate(prioritized_tasks.split("\n"))}
                self.task_list.sort(key=lambda x: priority_dict.get(x["name"], float("inf")))
            
            # Store tasks in vector store for future reference
            if execution_result:
                self.vectorstore.add_texts(
                    texts=[execution_result],
                    metadatas=[{"task": current_task["name"]}]
                )
        
        return {
            "objective": objective,
            "completed_tasks": self.completed_tasks,
            "results": results,
            "remaining_tasks": self.task_list
        }

    async def cleanup(self):
        """Clean up resources used by the BabyAGI agent."""
        await super().cleanup()  # Call parent cleanup
        if hasattr(self, 'vectorstore'):
            self.vectorstore.delete_collection()
        self.task_list = []
        self.completed_tasks = []
        self.task_id_counter = 0 